File: backtracking/46.py
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        n = len(nums)
        ret = []
        def perm(nums, k):
            if k == 1:
                ret.append(copy.copy(nums))
                return
            for i in range(0,k):
                #注意n-k与n-k+i的区别 标出每次交换的位置不同
                nums[n-k], nums[n-k+i] = nums[n-k+i], nums[n-k]   
                perm(nums, k-1)
                nums[n-k], nums[n-k+i] = nums[n-k+i], nums[n-k]
            
        perm(nums, n)
        return ret
            
#dfs
class Solution:
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        ret = []
        self.dfs(ret, [], nums)
        return ret
        
    def dfs(self, ret, temp, nums):
        if len(temp) == len(nums):
            ret.append(temp)
        for i in range(0, len(nums)):
            if nums[i] in temp:
                continue
            else:
                self.dfs(ret, temp+[nums[i]], nums)
        
            
            
             
        

File: backtracking/46.go
func permute(nums []int) [][]int {
    if len(nums) == 0 {
        return [][]int{}
    }
    ret := make([][]int, 0)
    bt(&ret, nums, []int{})
    return ret
}

func bt(ret *[][]int, nums []int, cur []int) {
    if len(nums) == 0 {
        // temp := append([]int{}, cur...)
        *ret = append(*ret, append([]int{}, cur...))
    }else {
        for i := 0; i < len(nums); i++ {
            temp := append([]int{}, nums[:i]...)
            tempcur := append([]int{}, cur...)
            // nums剔除当前位置元素, 递归剩下元素的permute,  cur加上当前i位置元素, go的复制操作有点日狗
            bt(ret, append(temp, nums[i+1:]...), append(tempcur, nums[i]))
        }
    }
}
